typed-protocols/src/Network/TypedProtocol/Driver.hs

```diff
@@ -282,24 +306,23 @@ runPipelinedPeerReceiverQueue
   :: forall ps pr failure bytes m c.
      (MonadSTM m, MonadThrow m, Exception failure)
   => Tracer m (TraceSendRecv ps)
-  -> TBQueue m (ReceiveHandler ps pr m c)
-  -> TBQueue m c
+  -> TBQueue m (ReceiveHandler bytes ps pr m c)
+  -> TBQueue m (c, Maybe bytes)
   -> Codec ps failure m bytes
   -> Channel m bytes
   -> m Void
 runPipelinedPeerReceiverQueue tr receiveQueue collectQueue codec channel =
     go Nothing
   where
     go :: Maybe bytes -> m Void
-    go trailing = do
-      ReceiveHandler n receiver <- atomically (readTBQueue receiveQueue)
-      case (n, trailing) of
-        -- This is the first pipelined response but we already had trailing
-        -- data! This cannot be legal.
-        (Zero, Just _) -> throwM IllegalTrailingData
-        _              -> return ()
+    go receiverTrailing = do
+      ReceiveHandler senderTrailing receiver
+        <- atomically (readTBQueue receiveQueue)
```

Comment:
This is the tricky part, the sender could place multiple `MaybeTrailingBytes ...`, but it won't since `go _ _ (SenderPipeline _)` will continue with `NoTrailing`; Next time the `MaybeTrailingBytes` will be seen in the receiver is when we collect all responses and start pipelinging again.  And nice thing about this approach is that this is enforced by the types, so it's not just a runtime property.